File: RoombAPI.Server/Services/SchedulerService.cs
namespace RoombaPi.Server.Services
{
  using System;
  using System.Linq;
  using System.Threading.Tasks;
  using Create2.Schedule;
  using global::Grpc.Core;
  using Google.Protobuf.WellKnownTypes;
  using RoombAPI.Components;
  using RoombaPi.Grpc;
  using Scheduler = RoombaPi.Grpc.Scheduler;

  public class SchedulerService : Scheduler.SchedulerBase
  {
    private readonly RoombaScheduler scheduler;

    public SchedulerService(RoombaScheduler scheduler)
    {
      this.scheduler = scheduler;
    }

    public override async Task<Empty> ClearSchedule(Empty request, ServerCallContext context)
    {
      await this.scheduler.ClearSchedule();
      return new Empty();
    }

    public override Task<GetClockResponse> GetClock(Empty request, ServerCallContext context)
    {
      var clockTime = this.scheduler.GetClockTime();
      return Task.FromResult(new GetClockResponse
      {
        Day = this.ConvertToGrpcWeekDay(clockTime.Day),
        Hour = clockTime.Hour,
        Minute = clockTime.Minute
      });
    }

    public override Task<GetScheduleResponse> GetSchedule(Empty request, ServerCallContext context)
    {
      var schedule = this.scheduler.GetSchedule().Select(s => new ScheduleDay
        { Day = this.ConvertToGrpcWeekDay(s.Day), Hour = s.Hour, Minute = s.Minute });
      return Task.FromResult(new GetScheduleResponse { Days = { schedule } });
    }

    public override async Task<Empty> SetSchedule(SetScheduleRequest request, ServerCallContext context)
    {
      await this.scheduler.SetSchedule(request.Days.Select(d => new DayTime
        {Day = this.ConvertToCreate2WeekDay(d.Day), Hour = (byte)d.Hour, Minute = (byte)d.Minute}).ToArray());
      return new Empty();
    }

    private WeekDay ConvertToGrpcWeekDay(WeekDays day)
    {
      switch (day)
      {
        case WeekDays.Sunday:
          return WeekDay.Sunday;
        case WeekDays.Monday:
          return WeekDay.Monday;
        case WeekDays.Tuesday:
          return WeekDay.Tuesday;
        case WeekDays.Wednesday:
          return WeekDay.Wednesday;
        case WeekDays.Thursday:
          return WeekDay.Thursday;
        case WeekDays.Friday:
          return WeekDay.Friday;
        case WeekDays.Saturday:
          return WeekDay.Saturday;
        default:
          throw new ArgumentOutOfRangeException(nameof(day), day, null);
      }
    }

    private WeekDays ConvertToCreate2WeekDay(WeekDay day)
    {
      return day switch
      {
        WeekDay.Sunday => WeekDays.Sunday,
        WeekDay.Monday => WeekDays.Monday,
        WeekDay.Tuesday => WeekDays.Tuesday,
        WeekDay.Wednesday => WeekDays.Wednesday,
        WeekDay.Thursday => WeekDays.Thursday,
        WeekDay.Friday => WeekDays.Friday,
        WeekDay.Saturday => WeekDays.Saturday,
        _ => throw new ArgumentOutOfRangeException(nameof(day), day, null)
      };
    }
  }
}

File: RoombaPi.Client/src/app/scheduler/services/scheduler.service.ts
import { Injectable } from "@angular/core";

import { environment } from "src/environments/environment";

import {
  SchedulerServiceClient,
  SetClockRequest,
  SetScheduleRequest,
  ScheduleDay as RequestScheduleDay
} from "@app/grpc/scheduler";
import { Empty } from "google-protobuf/google/protobuf/empty_pb";

import { WeekDay } from "../models/weekDay";
import { ScheduleDay } from "../models/scheduleDay";

@Injectable({ providedIn: "root" })
export class SchedulerService {
  private readonly client: SchedulerServiceClient;

  constructor() {
    this.client = new SchedulerServiceClient(environment.grpcServer);
  }

  public setClock(day: WeekDay, hour: number, minute: number): void {
    const request = new SetClockRequest();
    request.setDay(day);
    request.setHour(hour);
    request.setMinute(minute);
    this.client.setClock(request, null);
  }

  public SetSchedule(days: Array<ScheduleDay>): void {
    const request = new SetScheduleRequest();
    days.forEach(x => {
      const day = new RequestScheduleDay();
      day.setDay(x.day);
      day.setHour(x.hour);
      day.setMinute(x.minute);
      request.addDays(day);
    });

    this.client.setSchedule(request, null);
  }

  public clearSchedule(): void {
    this.client.clearSchedule(new Empty(), null);
  }
}
